File: src/main/java/org/teamvoided/transition/utils/RegionFileIO.java
package org.teamvoided.transition.utils;

import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.NbtIo;
import net.minecraft.world.level.ChunkPos;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.chunk.storage.RegionFile;
import net.minecraft.world.level.chunk.storage.RegionStorageInfo;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.File;
import java.io.IOException;
import java.nio.file.Path;
import java.util.HashMap;

import static org.teamvoided.transition.Transition.LOGGER;
import static org.teamvoided.transition.Transition.log;

public interface RegionFileIO {
    RegionStorageInfo fakeData = new RegionStorageInfo("", Level.OVERWORLD, "");
    static RegionFile getRegionFile(Path file) throws IOException {
        return new RegionFile(fakeData, file, file.getParent(), false);
    }

    int REGION_CHUNKS = 32;

    static HashMap<ChunkPos, CompoundTag> read(File file) throws IOException {
        RegionFile regionFile = getRegionFile(file.toPath());

        // r.0.0.mca
        var cordArray = file.getName().split("\\.");

        var x = Integer.parseInt(cordArray[1]);
        var z = Integer.parseInt(cordArray[2]);

        var map = new HashMap<ChunkPos, CompoundTag>();
                LOGGER.error("REGION FILE READER IS STILL WRONG!!!");
        for (int i = 0; i < REGION_CHUNKS; i++) {
            for (int j = 0; j < REGION_CHUNKS; j++) {
                var chunkPos = new ChunkPos((x * REGION_CHUNKS) + i, (z * REGION_CHUNKS) + j);
                if (regionFile.doesChunkExist(chunkPos)) map.put(chunkPos, read(regionFile, chunkPos));
            }
        }
        return map;

    }

    static CompoundTag read(RegionFile regionFile, ChunkPos chunkPos) throws IOException {
        DataInputStream dataInputStream = regionFile.getChunkDataInputStream(chunkPos);
        CompoundTag tag;
        funny:
        {
            try {
                if (dataInputStream == null) {
                    tag = null;
                    break funny;
                }

                tag = NbtIo.read(dataInputStream);
            } catch (Throwable var7) {
                try {
                    dataInputStream.close();
                } catch (Throwable var6) {
                    var7.addSuppressed(var6);
                }

                throw var7;
            }
            dataInputStream.close();
            return tag;
        }

        if (dataInputStream != null) {
            dataInputStream.close();
        }

        return tag;
    }

    static void write(File file, ChunkPos chunkPos, CompoundTag compoundTag) throws IOException {
        write(getRegionFile(file.toPath()), chunkPos, compoundTag);
    }

    static void write(RegionFile regionFile, ChunkPos chunkPos, CompoundTag compoundTag) throws IOException {
        DataOutputStream dataOutputStream = regionFile.getChunkDataOutputStream(chunkPos);
        try {
            NbtIo.write(compoundTag, dataOutputStream);
        } catch (Throwable var8) {
            if (dataOutputStream != null) {
                try {
                    dataOutputStream.close();
                } catch (Throwable var7) {
                    var8.addSuppressed(var7);
                }
            }

            throw var8;
        }
        dataOutputStream.close();
    }
}

File: src/main/kotlin/org/teamvoided/transition/utils/RegionFileIO.kt
package org.teamvoided.transition.utils

import net.minecraft.core.registries.Registries
import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.NbtIo
import net.minecraft.resources.ResourceKey
import net.minecraft.world.level.ChunkPos
import net.minecraft.world.level.Level
import net.minecraft.world.level.chunk.storage.RegionFile
import net.minecraft.world.level.chunk.storage.RegionStorageInfo
import org.teamvoided.transition.Transition
import org.teamvoided.transition.Transition.MODID
import org.teamvoided.transition.Transition.id
import java.io.File
import java.io.IOException
import java.nio.file.Path

object RegionFileIO {
    const val REGION_CHUNKS: Int = 32
    val fakeDimension: ResourceKey<Level> = ResourceKey.create(Registries.DIMENSION, id("none"))
    val fakeData = RegionStorageInfo("${MODID}_world", fakeDimension, MODID)

    @Throws(IOException::class)
    fun getRegionFile(file: Path): RegionFile = RegionFile(fakeData, file, file.parent, false)

    @Throws(IOException::class)
    fun read(file: File): HashMap<ChunkPos, CompoundTag> {
        val regionFile = getRegionFile(file.toPath())

        // r.0.0.mca
        val cordArray = file.getName().split("\\.".toRegex()).dropLastWhile { it.isEmpty() }

        val x = cordArray[1].toInt()
        val z = cordArray[2].toInt()

        val map = HashMap<ChunkPos, CompoundTag>()
        for (i in 0..<REGION_CHUNKS) {
            for (j in 0..<REGION_CHUNKS) {
                val chunkPos = ChunkPos((x * REGION_CHUNKS) + i, (z * REGION_CHUNKS) + j)
                if (regionFile.doesChunkExist(chunkPos)) {
                    try {
                        val nbt = read(regionFile, chunkPos)
                            ?: throw NullPointerException("Data input stream was null in region: ${file.name}, chunk: $chunkPos")
                        map.put(chunkPos, nbt)
                    } catch (e: Throwable) {
                        Transition.error("Failed to read chunk: $chunkPos in region: ${file.name}")
                        Transition.error(e.message)
                        Transition.error(e.stackTraceToString())
                    }
                }
            }
        }
        return map
    }

    @Throws(IOException::class)
    fun read(regionFile: RegionFile, chunkPos: ChunkPos): CompoundTag? {
        val dataInputStream = regionFile.getChunkDataInputStream(chunkPos) ?: return null
        var tag: CompoundTag
        try {
            tag = NbtIo.read(dataInputStream)
        } catch (var7: Throwable) {
            try {
                dataInputStream.close()
            } catch (var6: Throwable) {
                var7.addSuppressed(var6)
            }
            throw var7
        }
        try {
            dataInputStream.close()
        } catch (e: Throwable) {
            throw e
        }
        return tag
    }

    @Throws(IOException::class)
    fun write(file: File, chunkPos: ChunkPos, compoundTag: CompoundTag) {
        write(getRegionFile(file.toPath()), chunkPos, compoundTag)
    }

    @Throws(IOException::class)
    fun write(regionFile: RegionFile, chunkPos: ChunkPos, compoundTag: CompoundTag) {
        val dataOutputStream = regionFile.getChunkDataOutputStream(chunkPos)
        try {
            NbtIo.write(compoundTag, dataOutputStream)
        } catch (var8: Throwable) {
            if (dataOutputStream != null) {
                try {
                    dataOutputStream.close()
                } catch (var7: Throwable) {
                    var8.addSuppressed(var7)
                }
            }

            throw var8
        }
        dataOutputStream.close()
    }
}